### PR TITLE
ros2log new package introduction

### PR DESCRIPTION
**Pull request by fujitatomoya**
## Description

part of https://github.com/ros2/ros2cli/issues/1148 for `Node Granurality`

Fixes # (issue)

### Is this user-facing behavior change?

Yes, new package is introduced `ros2log`

### Did you use Generative AI?

Yes we did during the development. Claude Sonnet 4.6.

### CODE REVIEW

**Comment by fujitatomoya** [comment]
CC: @decwest

**Review by fujitatomoya — COMMENTED** [review]
@Barry-Xu-2018 

this is major enhancement for ros2log command that has been developed by ROS Japan Dev team. can you take a 1st look and review this?

**Comment by fujitatomoya** [comment]
@decwest i will start the CI for ros2log. this is gonna be the 1st time for other platforms like RHEL and window, i would expect that there are going to be several fixes to pass.

**Comment by fujitatomoya** [comment]
Pulls: ros2/ros2cli#1217
Gist: https://gist.githubusercontent.com/fujitatomoya/56b5badffe7e633cceecafd22bf7230f/raw/2aa772334eb4fd9e80c78070cc7708c9e3e908dd/ros2.repos
BUILD args:  --packages-up-to ros2log
TEST args:  --packages-select ros2log
ROS Distro: humble
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18692
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/28366/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/21242/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/8226/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/27244/)

**Comment by fujitatomoya** [comment]
Pulls: ros2/ros2cli#1217
Gist: https://gist.githubusercontent.com/fujitatomoya/d4d623a2ac8a7ba8a901274ee598172e/raw/6e539e43c41889088e02ba07135dd567298f1f70/ros2.repos
BUILD args:  --packages-up-to ros2log
TEST args:  --packages-select ros2log
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18694
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/28368/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/21244/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/8228/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/27246/)

**Comment by fujitatomoya** [comment]
Pulls: ros2/ros2cli#1217
Gist: https://gist.githubusercontent.com/fujitatomoya/8c77c69338fdae4ee6dae5eb8a61f8e1/raw/6e539e43c41889088e02ba07135dd567298f1f70/ros2.repos
BUILD args:  --packages-up-to ros2log
TEST args:  --packages-select ros2log
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18695
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/28369/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/21245/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/8229/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/27247/)

File: ros2log/ros2log/verb/__init__.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2025 Tomoya Fujita, Fumiya Ohnishi
```
Comment (Barry-Xu-2018) [inline ros2log/ros2log/verb/__init__.py]: The year is incorrect.  
There are other files with this issue as well.

File: ros2log/ros2log/api/__init__.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+    while futures and not all(future.done() for future in futures.values()):
+        rclpy.spin_once(node, timeout_sec=0.1)
+
+    for node_name, future in futures.items():
+        if future.result() is not None:
+            results[node_name] = list(future.result().levels)
+        else:
+            results[node_name] = future.exception()
+
```
Comment (Barry-Xu-2018) [inline ros2log/ros2log/api/__init__.py:176]: Before returning from the function `call_get_logger_levels`, all created clients should be destroyed by node.destroy_client().

File: ros2log/ros2log/api/__init__.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    while futures and not all(future.done() for future in futures.values()):
+        rclpy.spin_once(node, timeout_sec=0.1)
+
+    for node_name, future in futures.items():
+        if future.result() is not None:
+            results[node_name] = list(future.result().results)
+        else:
+            results[node_name] = future.exception()
+
```
Comment (Barry-Xu-2018) [inline ros2log/ros2log/api/__init__.py]: Before returning from the function `call_set_logger_levels`, all created clients should be destroyed by node.destroy_client().

File: ros2log/package.xml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  <maintainer email="tomoya.fujita825@gmail.com">Tomoya Fujita</maintainer>
+  <maintainer email="fumiya-onishi@keio.jp">Fumiya Ohnishi</maintainer>
+  <license>Apache License 2.0</license>
+
+  <author email="tomoya.fujita825@gmail.com">Tomoya Fujita</author>
+  <author email="fumiya-onishi@keio.jp">Fumiya Ohnishi</author>
+
+  <exec_depend>python3-argcomplete</exec_depend>
+  <exec_depend>python3-packaging</exec_depend>
+  <exec_depend>python3-psutil</exec_depend>
```
Comment (Barry-Xu-2018) [inline ros2log/package.xml]: These two packages aren't used in the code, so they can be removed from the dependencies.

File: ros2log/ros2log/verb/__init__.py (hunk 1 above)
Comment (decwest) [reply ros2log/ros2log/verb/__init__.py]: Thank you for your feedback. I have addressed the issue in the following commit: https://github.com/ros2/ros2cli/pull/1217/changes/0ee0e4d99736ce1dd8fb61191a586798ccf6d1af

File: ros2log/ros2log/api/__init__.py (hunk 2 above)
Comment (decwest) [reply ros2log/ros2log/api/__init__.py:176]: Understood. I have addressed this in the following commit: https://github.com/ros2/ros2cli/pull/1217/changes/934503e4efa262e7ad3a0caffb95ac79d8c24d52

File: ros2log/ros2log/api/__init__.py (hunk 3 above)
Comment (decwest) [reply ros2log/ros2log/api/__init__.py]: Ditto

File: ros2log/package.xml (hunk 4 above)
Comment (decwest) [reply ros2log/package.xml]: Thank you for your feedback. I have addressed the issue in the following commit: https://github.com/ros2/ros2cli/pull/1217/changes/10cdadb3a52660226e4f568670ca9df28fd1aa07

**Comment by decwest** [comment]
@fujitatomoya
Regarding the CI build failure, the following report suggests that the cause is `ModuleNotFoundError: No module named 'ros2node'`.
https://ci.ros2.org/job/ci_linux/28369/testReport/%28root%29/ros2log/___/

To address this, I added `ros2node` as an `exec_depend` in the following commit.
https://github.com/ros2/ros2cli/pull/1217/changes/2254ca5ce4b68daad4dfaff4c4dda9e45dbb7faf

Also, since `test_cli.py` imports launch-related modules, I added the corresponding `test_depend` entries as well.

File: ros2log/ros2log/verb/__init__.py (hunk 1 above)
Comment (fujitatomoya) [reply ros2log/ros2log/verb/__init__.py]: when we started the development, that was 2025. so it should be fine not to change... resolving anyway since @decwest addressed this.

**Comment by fujitatomoya** [comment]
Pulls: ros2/ros2cli#1217
Gist: https://gist.githubusercontent.com/fujitatomoya/13bac608e7f568daece4dc4e371e52b5/raw/6e539e43c41889088e02ba07135dd567298f1f70/ros2.repos
BUILD args:  --packages-up-to ros2log
TEST args:  --packages-select ros2log
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/18736
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/28414/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/21286/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/8273/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/27291/)

**Review by Barry-Xu-2018 — APPROVED** [review]
LGTM

**Comment by decwest** [comment]
I addressed the following build error by this commit: https://github.com/ros2/ros2cli/pull/1217/commits/d5bdf970aaf627fd4166dcab3f50bca8d0bf4742

https://ci.ros2.org/job/ci_linux-rhel/8273/testReport/(root)/ros2log/___/

**Comment by fujitatomoya** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/28469/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/21333/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/8321/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/27337/)

File: ros2log/ros2log/api/__init__.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+from typing import Optional
+from typing import Union
+
+from rcl_interfaces.msg import LoggerLevel
+from rcl_interfaces.msg import SetLoggerLevelsResult
+from rcl_interfaces.srv import GetLoggerLevels
+from rcl_interfaces.srv import SetLoggerLevels
+import rclpy
+
+from ros2node.api import get_absolute_node_name
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py:27]:
```suggestion
from rclpy.node import Node

```

File: ros2log/ros2log/api/__init__.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    absolute_node_name = _require_absolute_node_name(node_name)
+    return (
+        f"Logger service not available for node '{absolute_node_name}'.\n"
+        "The 'enable_logger_service' node option must be enabled for this node."
+    )
+
+
+def iter_logger_service_nodes(
+    *,
+    node,
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py]:
```suggestion
    node: Node,
```

File: ros2log/ros2log/api/__init__.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+) -> Iterator[NodeName]:
+    """Yield nodes that expose the logger get/set services as they are discovered."""
+    for node_name in get_node_names(node=node):
+        if node_has_logger_services(node, node_name):
+            yield node_name
+
+
+def get_logger_service_nodes(*, node) -> list[NodeName]:
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py]:
```suggestion
def get_logger_service_nodes(*, node: Node) -> list[NodeName]:
```

File: ros2log/ros2log/api/__init__.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    """Return all nodes that expose the logger get/set services."""
+    return list(iter_logger_service_nodes(node=node))
+
+
+def get_target_node_names(
+    *,
+    node,
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py]:
```suggestion
    node: Node,
```

File: ros2log/ros2log/api/__init__.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    if absolute_node_name not in all_node_names:
+        return None, 'Node not found'
+
+    if absolute_node_name not in logger_service_node_names:
+        return None, format_logger_service_unavailable_error(absolute_node_name)
+
+    return [absolute_node_name], None
+
+
+def node_has_logger_services(node, node_name: NodeName) -> bool:
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py]:
```suggestion
def node_has_logger_services(node: Node, node_name: NodeName) -> bool:
```

File: ros2log/ros2log/api/__init__.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+    return (
+        _service_has_type(service_map, expected_get, LOGGER_GET_SERVICE_TYPE) and
+        _service_has_type(service_map, expected_set, LOGGER_SET_SERVICE_TYPE)
+    )
+
+
+def call_get_logger_levels(
+    *,
+    node,
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py]:
```suggestion
    node: Node,
```

File: ros2log/ros2log/api/__init__.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+        return results
+    finally:
+        for client in clients.values():
+            node.destroy_client(client)
+
+
+def call_set_logger_levels(
+    *,
+    node,
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py]:
```suggestion
    node: Node,
```

File: ros2log/ros2log/verb/watch.py [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+    def __init__(
+        self,
+        node: Node,
+        level_filter: Optional[str] = None,
+        logger_filter: Optional[str] = None,
+        regex_filter: Optional[str] = None,
+        enable_color: bool = True,
+        show_timestamp: bool = True,
+        show_function_detail: bool = False,
+        qos_profile=qos_profile_rosout_default,
```
Comment (ahcorde) [inline ros2log/ros2log/verb/watch.py]:
```suggestion
        qos_profile: QoSProfile = qos_profile_rosout_default,
```

File: ros2log/ros2log/verb/watch.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+
+import re
+import sys
+from typing import Optional
+
+from rcl_interfaces.msg import Log
+
+import rclpy
+from rclpy.node import Node
+from rclpy.qos import qos_profile_rosout_default
```
Comment (ahcorde) [inline ros2log/ros2log/verb/watch.py:23]:
```suggestion
from rclpy.qos import qos_profile_rosout_default
from rclpy.qos import QoSProfile
```

File: ros2log/ros2log/verb/watch.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        level_filter: Optional[str] = None,
+        logger_filter: Optional[str] = None,
+        regex_filter: Optional[str] = None,
+        enable_color: bool = True,
+        show_timestamp: bool = True,
+        show_function_detail: bool = False,
+        qos_profile=qos_profile_rosout_default,
+        enable_content_filter: bool = True,
+        debug: bool = False,
+    ):
```
Comment (ahcorde) [inline ros2log/ros2log/verb/watch.py]:
```suggestion
    ) -> None:
```

File: ros2log/ros2log/verb/watch.py [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+        self.logger_filter = logger_filter
+
+        # Set up regex filter
+        self.regex_pattern = None
+        if regex_filter:
+            try:
+                self.regex_pattern = re.compile(regex_filter)
+            except re.error as e:
+                node.get_logger().error(f'Invalid regex pattern: {e}')
+                sys.exit(1)
```
Comment (ahcorde) [inline ros2log/ros2log/verb/watch.py]: maybe raise here?

File: ros2log/ros2log/verb/watch.py [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+        add_qos_arguments(
+            parser,
+            entity_type='subscribe',
+            default_profile_str='rosout_default')
+
+    def main(self, *, args):
+        with DirectNode(args) as node:
+            # Configure QoS profile based on arguments and available publishers
+            qos_profile = choose_qos(node, '/rosout', args)
+            LogWatcher(
```
Comment (ahcorde) [inline ros2log/ros2log/verb/watch.py]: can try raise an exception?

File: ros2log/ros2log/api/__init__.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+
+
+def call_get_logger_levels(
+    *,
+    node,
+    logger_names_by_node: dict[str, Sequence[str]],
+    timeout_sec: float = 5.0,
+) -> dict[str, Union[list[LoggerLevel], Exception]]:
+    """Call the get-logger-levels service for one or more nodes."""
+    clients = {}
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py:146]:
```suggestion
    if timeout_sec <= 0:                                                                                                                                                           
      raise ValueError(f'timeout_sec must be positive, got {timeout_sec}')
    clients = {}
```

File: ros2log/ros2log/api/__init__.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+
+
+def call_set_logger_levels(
+    *,
+    node,
+    levels_by_node: dict[str, Sequence[LoggerLevel]],
+    timeout_sec: float = 5.0,
+) -> dict[str, Union[list[SetLoggerLevelsResult], Exception]]:
+    """Call the set-logger-levels service for one or more nodes."""
+    clients = {}
```
Comment (ahcorde) [inline ros2log/ros2log/api/__init__.py:191]: check here timeout_sec < 0 ?

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: ros2log/ros2log/api/__init__.py (hunk 18 above)
Comment (decwest) [reply ros2log/ros2log/api/__init__.py:191]: Thank you for the review. I have applied the fixes in this commit: https://github.com/ros2/ros2cli/pull/1217/changes/d55e3722c42cc5ef0d586748579992474fcc714c

I have also added tests to verify that a negative value cannot be specified for timeout_sec

File: ros2log/ros2log/verb/watch.py (hunk 16 above)
Comment (fujitatomoya) [reply ros2log/ros2log/verb/watch.py]: i agree with this.

the constructor should raise an exception (e.g., ValueError or re-raise the re.error). Constructors shouldn't call sys.exit(), it makes the class untestable and non-reusable.

Comment (fujitatomoya) [reply ros2log/ros2log/verb/watch.py]: https://github.com/ros2/ros2cli/pull/1217/commits/89b8337b304b63b32055e72d112006517393e040

File: ros2log/ros2log/verb/watch.py (hunk 15 above)
Comment (fujitatomoya) [reply ros2log/ros2log/verb/watch.py]: https://github.com/ros2/ros2cli/pull/1217/commits/89b8337b304b63b32055e72d112006517393e040

**Comment by fujitatomoya** [comment]
@decwest as you may know, everyone is busy with Lyrical Luth release (feature freeze, branching and so on), so it will be hard to get this in the mainline source code right at this moment. we could probably come back on this to get more review before merge after lyrical luth release.

in the mean time, i will start the CI for ros2log package. we can expect that there should be (will be) no conflicts since this is new whole package to the mainline.

**Comment by fujitatomoya** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/28693/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/21515/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/8532/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/27565/)

**Comment by fujitatomoya** [comment]
Pulls: ros2/ros2cli#1217
Gist: https://gist.githubusercontent.com/fujitatomoya/89644b550c3c5f9bdb19357b5435cae0/raw/f5ffcf545aa04234bb725f6f109580c971ecb4fe/ros2.repos
BUILD args:  --packages-up-to ros2log
TEST args:  --packages-select ros2log
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/19908
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/29856/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/22575/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/9767/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/28765/)

**Comment by fujitatomoya** [comment]
@ahcorde we have addressed all of your comments, can you approve?

@InvincibleRMC we would love to have 2nd review on this, can you review this? (even partial review is appreciated.)

**Comment by fujitatomoya** [comment]
Building on the built-in node in workspace /var/lib/jenkins/workspace/ci_launcher
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/29865/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/22584/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/9776/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/28774/)

**Review by ahcorde — APPROVED** [review]

File: ros2log/ros2log/api/__init__.py [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections.abc import Iterator
+from collections.abc import Sequence
+import time
+from typing import Callable
```
Comment (InvincibleRMC) [inline ros2log/ros2log/api/__init__.py]: This should be imported from collections.abc this is deprecated.

File: ros2log/ros2log/api/__init__.py [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+    """Return all nodes that expose the logger get/set services."""
+    return list(iter_logger_service_nodes(node=node))
+
+
+def get_target_node_names(
+    *,
+    node: Node,
+    node_name: Optional[str] = None,
+    all_nodes: bool = False,
+) -> tuple[Optional[list[str]], Optional[str]]:
```
Comment (InvincibleRMC) [inline ros2log/ros2log/api/__init__.py]: This is more percise

```suggestion
) -> tuple[list[str], None] | tuple[None, str]:
```

File: ros2log/ros2log/verb/levels.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+    'ERROR': (
+        'Error is for information that this node cannot resolve.'
+    ),
+    'FATAL': (
+        'Information about an impending node shutdown.'
+    ),
+}
+
+
+def get_log_levels_from_msg():
```
Comment (InvincibleRMC) [inline ros2log/ros2log/verb/levels.py]:
```suggestion
def get_log_levels_from_msg() -> list[tuple[str, int]]:
```

File: ros2log/ros2log/verb/watch.py [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+
+
+class WatchVerb(VerbExtension):
+    """Monitor and display logs in real-time."""
+
+    def add_arguments(self, parser, cli_name):
+        parser.add_argument(
+            '--level',
+            type=str,
+            choices=['DEBUG', 'INFO', 'WARN', 'ERROR', 'FATAL'],
```
Comment (InvincibleRMC) [inline ros2log/ros2log/verb/watch.py]:
```suggestion
            choices=LOG_LEVELS.keys(),
```

File: ros2log/test/fixtures/listener_node.py [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    def __init__(self):
+        super().__init__('listener', enable_logger_service=False)
+        self.subscription = self.create_subscription(
+            String,
+            'chatter',
+            self.listener_callback,
+            10)
+        self.get_logger().info('Listener node started')
+
+    def listener_callback(self, msg):
```
Comment (InvincibleRMC) [inline ros2log/test/fixtures/listener_node.py]:
```suggestion
    def listener_callback(self, msg: String) -> None:
```

File: ros2log/package.xml [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+  <exec_depend>rclpy</exec_depend>
+  <exec_depend>ros2cli</exec_depend>
+  <exec_depend>ros2node</exec_depend>
+  <exec_depend>rosgraph_msgs</exec_depend>
+
+  <test_depend>ament_copyright</test_depend>
+  <test_depend>ament_flake8</test_depend>
+  <test_depend>ament_pep257</test_depend>
+  <test_depend>ament_xmllint</test_depend>
+  <test_depend>launch</test_depend>
```
Comment (InvincibleRMC) [inline ros2log/package.xml:25]: @fujitatomoya thoughts on adding ament_mypy also?

**Review by InvincibleRMC — APPROVED** [review]
Approved with some minor comments

File: ros2log/test/fixtures/listener_node.py (hunk 23 above)
Comment (fujitatomoya) [reply ros2log/test/fixtures/listener_node.py]: fixed in https://github.com/ros2/ros2cli/pull/1217/commits/096f0f2d8fa1135be61a21c244fe5cbc1091ee0a

File: ros2log/ros2log/verb/watch.py (hunk 22 above)
Comment (fujitatomoya) [reply ros2log/ros2log/verb/watch.py]: fixed in https://github.com/ros2/ros2cli/pull/1217/commits/096f0f2d8fa1135be61a21c244fe5cbc1091ee0a

File: ros2log/ros2log/verb/levels.py (hunk 21 above)
Comment (fujitatomoya) [reply ros2log/ros2log/verb/levels.py]: fixed in https://github.com/ros2/ros2cli/pull/1217/commits/096f0f2d8fa1135be61a21c244fe5cbc1091ee0a

File: ros2log/ros2log/api/__init__.py (hunk 20 above)
Comment (fujitatomoya) [reply ros2log/ros2log/api/__init__.py]: fixed in https://github.com/ros2/ros2cli/pull/1217/commits/096f0f2d8fa1135be61a21c244fe5cbc1091ee0a

File: ros2log/ros2log/api/__init__.py (hunk 19 above)
Comment (fujitatomoya) [reply ros2log/ros2log/api/__init__.py]: fixed in https://github.com/ros2/ros2cli/pull/1217/commits/096f0f2d8fa1135be61a21c244fe5cbc1091ee0a

File: ros2log/package.xml (hunk 24 above)
Comment (fujitatomoya) [reply ros2log/package.xml:25]: i say a separate PR is the right call, and I'd even say it's worth doing, just not here.

**Comment by fujitatomoya** [comment]
@ahcorde @InvincibleRMC thank you very much for reviewing this.

hopefully final CI with `--pacakges-select ros2log` because this is new pacakge and nobody else depends on it. (that means, we can backport this to downstream distros if we want.)

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/29870/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/22589/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/9782/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/28779/)

CC: @decwest
